pkg/controller/persistentvolume/state.go

```diff
@@ -0,0 +1,225 @@
+//     docs/design/persistent-volume-controller-pvc-flow.png
+
+// Pending volumes are either not bound or require provisioning.
+type persistentVolumePendingState struct{}
+
+// Bound volumes have a ClaimRef and the volume is fully provisioned
+type persistentVolumeBoundState struct{}
+
+// Released volumes have a ClaimRef, but the claim 404s in the API. User deleted the claim.
+// 4 possible outcomes from this phase:
+//    1.  Retain - PV is kept in Released phase and is not available for another claim
+//	  2.  Recycle - PV is scrubbed clean and the ClaimRef is removed (will match Pending above on sync)
+//	  3.  Delete - PV is removed from infrastructure and PV is deleted via API
+//	  4.  Failed - PV failed to Recycle or Delete.
+type persistentVolumeReleasedState struct{}
+
+// persistentVolumePendingState determines if a PersistentVolume should be Pending phase.
+// The PV is Pending when it is not bound (no ClaimRef) or it is bound (has ClaimRef) but is not fully provisioned.
+func (state *persistentVolumePendingState) isCurrentState(ctx *persistentVolumeStateContext) bool {
+	if ctx.volume.Spec.ClaimRef == nil {
+		return true
+	}
+	if isBeingProvisioned(ctx.volume) {
+		return true
+	}
+	return false
+}
+
+func (state *persistentVolumePendingState) processState(ctx *persistentVolumeStateContext) error {
+	if err := ctx.updatePersistentVolumeStatus(api.VolumePending); err != nil {
+		return err
+	}
+
+	// launch Provisioner in separate thread / Goroutine
+	// lock on pv.name like the main control, guaranteeing 1 thread processing
+	// a PV at a time but allowing concurrent processing of multiple volumes
+	// TODO: make this work :)
+	if !isProvisioningComplete(ctx.volume) {
+		// go provisionVolume(ctx.volume)
+	}
+
+	return nil
+}
+
+// persistentVolumeBoundState determines if a PersistentVolume should be in Bound phase.
+// The PV is Pending when it is not bound (no ClaimRef) or it is bound (has ClaimRef) but is not fully provisioned.
+func (state *persistentVolumeBoundState) isCurrentState(ctx *persistentVolumeStateContext) bool {
+	if ctx.volume.Spec.ClaimRef == nil {
+		return false
+	}
+	if isBeingProvisioned(ctx.volume) {
+		return false
+	}
+	// have a claimRef but 404 means user deleted the claim. no longer bound, needs recycling.
+	if ctx.claimNotFound {
+		return false
+	}
+	// have ClaimRef and completed (or don't need) provisioning
+	return true
+}
+
+func (state *persistentVolumeBoundState) processState(ctx *persistentVolumeStateContext) error {
+	if err := ctx.updatePersistentVolumeStatus(api.VolumeBound); err != nil {
+		return err
+	}
+	return nil
+}
+
+// persistentVolumeReleasedState determines if a PersistentVolume should be in Released phase.
+func (state *persistentVolumeReleasedState) isCurrentState(ctx *persistentVolumeStateContext) bool {
+	// have claimRef but claim 404'd, indicating user deleted it, unbinding this volume
+	if ctx.volume.Spec.ClaimRef != nil && ctx.claimNotFound {
```

Comment:
There are 2 options:

1.  Add more `if` statements to the state objects and enforce only 1 can match at a time.
2.  Enforce the order of state matching, first one to match is the next operation.

I prefer the second approach because having a well-defined lifecycle (in the image) is helpful and would be less complex by removing the need for more `if` conditionals in various objects.  The first approach requires both more `if` statements and code to enforce the "only one" paradigm.

So, to me, less code and a well-defined lifecycle trumps more code and more branches in the diagram.